Example of classification and regression (#110)
GRAAL-Research/poutyne: docs/source/examples/classification_and_regression.rst

Regarding the complexity of the problem and the number of training/valid samples, we have a huge number of training/validation images. Since there are not a considerable variation between images (e.g., the eye coordinates in images do not vary considerably), using all images in the dataset is not necessary and will only increase the training time. Hence, we can seperate and use a portion of data as below:

.. code-block:: python

    train_subset = Subset(train_dataset, np.arange(1, 50000))
    valid_subset = Subset(valid_dataset, np.arange(1, 2000))
    test_subset = Subset(test_dataset, np.arange(1, 1000))
    train_dataloader = DataLoader(train_subset, batch_size=batch_size, shuffle=True)
    valid_dataloader = DataLoader(valid_subset, batch_size=batch_size, shuffle=False)
    test_dataloader = DataLoader(test_subset, batch_size=batch_size, shuffle=False)

Here, we can see an example from the training dataset. It shows an image of a person, printing the gender and also showing the location of the eyes. It is worth mentioning that as we resize the image, the coordinates of the eyes should also be changed with same ratio.

.. code-block:: python

>     sample_number = 189

Reviewer: The sample number is not the same in the notebook.

    image = train_dataset[sample_number][0]
    image = image.permute(1, 2, 0).detach().numpy()
    image_rgb = cv2.cvtColor(np.float32(image), cv2.COLOR_BGR2RGB)
    image_rgb = image_rgb * imagenet_std + imagenet_mean
    gender = "male" if int(train_dataset[sample_number][1][0][gender_index]) == 1 else "female"
    print("Gender is:", gender)
    w, h = 218, 178

>     (x_L, y_L) = train_dataset[sample_number][1][1][
>         0:2
>     ]  # The coordinates vector of the datasets starts with X_L, y_L, X_R, y_R

Reviewer: ```suggestion
    # The coordinates vector of the datasets starts with X_L, y_L, X_R, y_R
    (x_L, y_L) = train_dataset[sample_number][1][1][0:2]
```
To change in the notebook as well.

    (x_R, y_R) = train_dataset[sample_number][1][1][2:4]
    w_scale = image_size / w
    h_scale = image_size / h
    x_L, x_R = (x_L * h_scale), (x_R * h_scale)  # rescaling for the size of (224,224) and finaly to the range of [0,1]
    y_L, y_R = (y_L * w_scale), (y_R * w_scale)
    x_L, x_R = int(x_L), int(x_R)
    y_L, y_R = int(y_L), int(y_R)
    image_rgb = cv2.drawMarker(image_rgb, (x_L, y_L), (0, 255, 0))
    image_rgb = cv2.drawMarker(image_rgb, (x_R, y_R), (0, 255, 0))
    image_rgb = cv2.cvtColor(np.float32(image_rgb), cv2.COLOR_BGR2RGB)
    image_rgb = np.clip(image_rgb, 0, 1)
    plt.imshow(image_rgb)
    plt.axis("off")
    plt.show()

> .. image:: /_static/img/classification_and_regression/dataset_sample.png

Reviewer: The image here does not fit what appears in the notebook, even if we use the same sample number.

            x = self.model(x)  # extract features from CNN
            scores = x[:, : self.num_classes]  # class scores
            coords = x[:, self.num_classes :]  # coordinates
            return scores, torch.sigmoid(coords)  # sigmoid output is in the range of [0, 1]

.. code-block:: python

    network = ClassifierLocalizer(model_name='resnet18')

Loss function
=============

As we discussed before, we have two different tasks in this example. These tasks need different loss functions; Cross-Entropy loss for the classification and Mean Square Error loss for the regression. Below, we define a new loss function class that sums both losses to considers them simultaneously. However, as the regression is relatively a simpler task here (due to similarity of coordinates in the images), we apply a lower weight to MSEloss.

.. code-block:: python

    class ClassificationRegressionLoss(nn.Module):
        def __init__(self, W):
            super(ClassificationRegressionLoss, self).__init__()
            self.ce_loss = nn.CrossEntropyLoss()  # size_average=False
            self.mse_loss = nn.MSELoss()
            self.W = W

        def forward(self, y_pred, y_true):

>             # print(y_true[0][:,20])

Reviewer: ```suggestion
```
Remove this comment. To fix in the notebook as well

>             loss_cls = self.ce_loss(y_pred[0], y_true[0][:, 20])  # Cross Entropy Error (for classification)

Reviewer: ```suggestion
            loss_cls = self.ce_loss(y_pred[0], y_true[0][:, gender_index])  # Cross Entropy Error (for classification)
```
Use gender_index here. To fix in the notebook as well

.. role:: hidden
    :class: hidden-section

Gender Classification and Eyes Location Detection: A Two Task Problem
*********************************************************************

.. note::

    - See the notebook `here <https://github.com/GRAAL-Research/poutyne/blob/master/examples/classification_and_regression.ipynb>`_
    - Run in `Google Colab <https://colab.research.google.com/github/GRAAL-Research/poutyne/blob/master/examples/classification_and_regression.ipynb>`_

> In this example, we are going to implement a multi-task problem. We try to identify the gender of the people, as well as locating their eyes in the image. Hence, we have two different tasks: classification (to identify the gender) and regression (to find the location of the eyes). We are going to use a single network (A CNN) to perform both tasks, however, we will need to apply different loss functions, each proper to a specific task.

Reviewer: ```suggestion
In this example, we are going to implement a multi-task problem. We try to identify the gender of the people, as well as locating their eyes in the image. Hence, we have two different tasks: classification (to identify the gender) and regression (to find the location of the eyes). We are going to use a single network (a CNN) to perform both tasks, however, we will need to apply different loss functions, each proper to a specific task.
```

Let's import all the needed packages.

.. code-block:: python

    import math
    import os
    import matplotlib.pyplot as plt
    import numpy as np
    import pandas as pd
    import wget
    import zipfile
    import cv2
    import torch
    import torch.nn as nn
    import torch.nn.functional as F
    import torch.optim as optim

>     from torch.autograd import Variable

Reviewer: `Variable`s are deprecated and useless in recent versions of PyTorch. Thus, you can remove its uses everywhere and everything will work as expected.

Author: Sure. I used it o activate the gradient ( requires_grad=True). So is it also automatically done in the latest version?

Reviewer: In the places you used `requires_grad=True`, you don't need to.

    import torchvision.datasets as datasets
    import torchvision.models as models
    import torchvision.transforms as tfms
    from poutyne import set_seeds, Model, ModelCheckpoint, CSVLogger, Experiment, StepLR
    from torch.utils.data import DataLoader, Subset, Dataset
    from torchvision.utils import make_grid

Training Constants
==================

.. code-block:: python

    num_epochs = 15
    learning_rate = 0.001
    batch_size = 32
    image_size = 224
    valid_split_percent = 0.1
    w, h = 218, 178   # the width and the hight of original images before resizing
    momentum = 0.5
    set_seeds(42)
    imagenet_mean = [0.485, 0.456, 0.406]  # mean of the ImageNet dataset for normalizing 
    imagenet_std = [0.229, 0.224, 0.225]  # std of the ImageNet dataset for normalizing
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    print('The running processor is...', device) 

CelebA Dataset
==============

> The section below consists of a few lines of codes that help us download celebA dataset from a public web source and unzip them. Downloading the Celeba dataset can be also done directly by `torch.datasets.CelebA(data_root, download =  True)`. however, due to the high traffic of Google Drive ( the main source of the dataset) it usually fails to function. hence we decided to download it from another public source but use it with `torch.datasets.CelebA()`.

Reviewer: ```suggestion
The section below consists of a few lines of codes that help us download CelebA dataset from a public web source and unzip them. Downloading the CelebA dataset can be also done directly by `torch.datasets.CelebA(data_root, download=True)`. However, due to the high traffic on the dataset's Google Drive (the main source of the dataset), it usually fails to function. Hence we decided to download it from another public source but use it with `torch.datasets.CelebA()`.
```

Fetching data 
=============

> The section below consists of a few lines of codes that help us download celebA dataset from a public web source and unzip them.

Reviewer: This repeats what has just been said.

Author: I didn't get your point. Do you mean I have repeated it in another place?

Reviewer: Yeah, the text above also says that.

.. code-block:: python

    data_root = 'datasets'
    base_url = "https://graal.ift.ulaval.ca/public/celeba/"

    file_list = [
        "img_align_celeba.zip",
        "list_attr_celeba.txt", 
        "identity_CelebA.txt",
        "list_bbox_celeba.txt",
        "list_landmarks_align_celeba.txt",
        "list_eval_partition.txt"
    ]

    # Path to folder with the dataset
    dataset_folder = f'{data_root}/celeba'
    os.makedirs(dataset_folder, exist_ok=True)

    for file in file_list:
        url = f"{base_url}/{file}"

>         if not os.path.exists(f"{dataset_folder}/{file}"):
>         wget.download(url, f"{dataset_folder}/{file}")

Reviewer: ```suggestion
        if not os.path.exists(f"{dataset_folder}/{file}"):
            wget.download(url, f"{dataset_folder}/{file}")
```

Missing tab.

    with zipfile.ZipFile(f"{dataset_folder}/img_align_celeba.zip", 'r') as ziphandler:
        ziphandler.extractall(dataset_folder)
   
Now, as the dataset id downloaded, we can define our datasets and dataloaders in its original way.

.. code-block:: python

>     transforms = tfms.Compose ([
>                             tfms.Resize((image_size, image_size)),
>                             tfms.ToTensor(),
>                             tfms.Normalize(imagenet_mean, imagenet_std)
>     ])

Reviewer: ```suggestion
    transforms = tfms.Compose([
        tfms.Resize((image_size, image_size)),
        tfms.ToTensor(),
        tfms.Normalize(imagenet_mean, imagenet_std)
    ])
```
Too many spaces.

                                    split='valid', 
                                    target_type=['attr', 'landmarks'], 
                                    transform=transforms)
    test_dataset = datasets.CelebA(data_root, 
                                split='test', 
                                target_type=['attr', 'landmarks'], 
                                transform=transforms)
                                
    train_dataloader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    valid_dataloader = DataLoader(valid_dataset, batch_size=batch_size, shuffle=False)
    test_dataloader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

Here we can see how each dataset sample looks like:

.. code-block:: python

    print (train_dataset[0])

Regarding the complexity of the problem and the high number training/valid samples, we can seperate and use a portion of data as below:

.. code-block:: python

    train_subset = Subset(train_dataset, np.arange(1,10000))
    valid_subset = Subset(valid_dataset, np.arange(1,2000))
    train_dataloader = DataLoader(train_subset, batch_size=batch_size, shuffle=True)

>     valid_dataloader = DataLoader(valid_subset, batch_size=batch_size, shuffle=False)

Reviewer: Not sure why we take a subset of the dataset. Is it because full training is too long?

Author: There is a huge number of images with a limited variation. (for example, eyes location don't vary very much). so using all images (189,643 images) is just time-consuming.

Reviewer: Alright, maybe note that in the text.

Here, we can see an example from the training dataset. It shows an image of a person, printing the gender and also showing the location of the eyes. It is worth mentioning that as we resize the image, the coordinates of the eyes should also be changed with same ratio.

.. code-block:: python

    sample_number = 189
    image = train_dataset[sample_number][0]
    image = image.permute(1,2,0).detach().numpy()
    image_rgb = cv2.cvtColor(np.float32(image), cv2.COLOR_BGR2RGB)
    image_rgb = image_rgb * imagenet_std + imagenet_mean

>     Gender = 'male' if int(train_dataset[sample_number][1][0][20])==1 else 'female'

Reviewer: I'm guessing that 20 is the index of the gender. I think it should be defined as a constant and explained after its definition. Otherwise, one might be confused by the code.

>     Gender = 'male' if int(train_dataset[sample_number][1][0][20])==1 else 'female'
>     print('Gender is: ', Gender)

Reviewer: ```suggestion
    gender = 'male' if int(train_dataset[sample_number][1][0][20])==1 else 'female'
    print('Gender is: ', gender)
```

    w, h = 218, 178
    (x_L, y_L) = train_dataset[sample_number][1][1][0:2]
    (x_R, y_R) = train_dataset[sample_number][1][1][2:4]
    w_scale = image_size/w
    h_scale = image_size/h
    x_L, x_R = (x_L*h_scale), (x_R*h_scale)  #rescaling for the size of (224,224) and finaly to the range of [0,1]
    y_L, y_R = (y_L*w_scale), (y_R*w_scale)
    x_L, x_R = int(x_L), int(x_R)
    y_L, y_R = int(y_L), int(y_R)
    image_rgb	= cv2.drawMarker(image_rgb, (x_L,y_L), (0,255,0))
    image_rgb	= cv2.drawMarker(image_rgb, (x_R,y_R), (0,255,0))
    image_rgb = cv2.cvtColor(np.float32(image_rgb), cv2.COLOR_BGR2RGB)
    image_rgb = np.clip(image_rgb ,0 , 1)
    plt.imshow(image_rgb)
    plt.axis('off')
    plt.show()

.. image:: /_static/img/classification_and_regression/dataset_sample.png

Network
=======

> Below, we define a new class, named 'ClassifierLocalizer, which accepts a pre-trained CNN and changes its last fully connected layer to be proper for the two task problem. The new fully connected layer contains 6 neurons, 2 for the classification task (male or female) and 4 for the localization task (x and y for the left and right eyes). Moreover, to put the location results on the same scale as the class scores, we apply the sigmoid function to the neurons assigned for the localization task.

Reviewer: ```suggestion
Below, we define a new class, named `ClassifierLocalizer`, which accepts a pre-trained CNN and changes its last fully connected layer to be proper for the two task problem. The new fully connected layer contains 6 neurons, 2 for the classification task (male or female) and 4 for the localization task (x and y for the left and right eyes). Moreover, to put the location results on the same scale as the class scores, we apply the sigmoid function to the neurons assigned for the localization task.
```

.. code-block:: python

    class ClassifierLocalizer(nn.Module):
        def __init__(self, model_name, num_classes=2):
            super(ClassifierLocalizer, self).__init__()
            self.num_classes = num_classes
            
            # create cnn model
            model = getattr(models, model_name)(pretrained=True)
            
            # remove fc layers and add a new fc layer
            num_features = model.fc.in_features
            model.fc = nn.Linear(num_features, 6) # classifier + localizer
            self.model = model
        
        def forward(self, x):
            x = self.model(x)                    # extract features from CNN
            scores = x[:, :self.num_classes]     # class scores
            coords = x[:, self.num_classes:]     # coordinates

>             return [scores, torch.sigmoid(coords)]   # sigmoid output is in the range of [0, 1]

Reviewer: ```suggestion
            return scores, torch.sigmoid(coords)   # sigmoid output is in the range of [0, 1]
```
Not need to return a list, a simple tuple will do.

> Regarding the complexity of the problem, the number of the samples in the training dataset, and the similarity of the training dataset to the ImageNet dataset, we may decide to freeze some of the layers. In our current example, based on the mentioned factors, we freeze just the last fully connected layer.

Reviewer: ```suggestion
Regarding the complexity of the problem, the number of the samples in the training dataset, and the similarity of the training dataset to the ImageNet dataset, we may decide to freeze some of the layers. In our current example, based on the mentioned factors, we freeze all layers but the last fully connected layer.
```

.. code-block:: python

    network = ClassifierLocalizer(model_name='resnet18')

    def freeze_weights(network):
        for name, param in network.named_parameters():
            if not name.startswith('fc.'):
                param.requires_grad = False

    freeze_weights(network)
    print(network)

.. code-block:: python

>     network = ClassifierLocalizer(model_name='resnet18')  # network without freezing any layer.

Reviewer: Not sure why you freeze the layers if you create after with no frozen layers.

Author: I will remove that, I was just examining it as a hyper parameter

Loss function
=============

> As we discussed before, we have two different tasks in this example. These tasks need different loss functions; Cross-Entropy loss for the classification and Mean Square Error loss for the regression. Below, we define a new loss function class that sums both losses to considers them simultaneously. However, as the regression is relatively a harder task, we apply a higher weight to MSEloss.

Reviewer: > we apply a higher weight to MSEloss.

Not sure where you apply this weight in the code below.

.. code-block:: python

    class ClassificationRegressionLoss(nn.Module):
        def __init__(self):
            super(ClassificationRegressionLoss, self).__init__()
            self.ce_loss = nn.CrossEntropyLoss() # size_average=False
            self.mse_loss = nn.MSELoss()
            
        def forward(self, y_pred, y_true):
            loss_cls = self.ce_loss(y_pred[0], Variable(y_true[0][:,20])) # Cross Entropy Error (for classification)
            loss_reg1 = self.mse_loss(y_pred[1][:,0], Variable(y_true[1][:,0]/h, requires_grad=True)) # Mean Squared Error for X_L
            loss_reg2 = self.mse_loss(y_pred[1][:,1], Variable(y_true[1][:,1]/w, requires_grad=True)) # Mean Squared Error for Y_L
            loss_reg3 = self.mse_loss(y_pred[1][:,2], Variable(y_true[1][:,2]/h, requires_grad=True)) # Mean Squared Error for X_R

>             loss_reg4 = self.mse_loss(y_pred[1][:,3], Variable(y_true[1][:,3]/w, requires_grad=True)) # Mean Squared Error for Y_R

Reviewer: This can be done in a single line, such as `loss_reg = self.mse_loss(y_pred[1], y_true[1])`.

Author: Yes if we just normalize the coordinates to another constant number. here I normalized them with their related coordinate length (h or w).

Reviewer: Oh right.

            total_loss = loss_cls + loss_reg1 + loss_reg2 + loss_reg3 + loss_reg4
            return total_loss

Training
========

.. code-block:: python

    optimizer = optim.Adam(network.parameters(), lr=learning_rate, weight_decay=0)
    loss_function = ClassificationRegressionLoss()
    #Step_Learning_Rate = StepLR(step_size=2 , gamma=0.1, last_epoch=-1, verbose=False)
    exp = Experiment('./two_task_example', network, optimizer=optimizer, loss_function=loss_function, device="all")
    exp.train(train_dataloader, valid_dataloader, callbacks=callbacks, epochs=num_epochs)

Evaluation
==========

> As you have also noticed from the training logs, we have achieved the best performance (considering the validation loss) at the 15th epoch. The weights of the network for the corresponding epoch have been automatically saved and we use these parameters to evaluate our algorithm visually. Hence,  we take advantage of evaluate function of Poutyne, and apply it to the validation dataset. It provides us the predictions as well as the ground-truth for comparison, in case of need.

Reviewer: > we have achieved the best performance (considering the validation loss) at the 15th epoch.

Is this always the case? Anyway, I think you could the load_checkpoint function in Experiment that allows to load the best checkpooint.

Author: No, i will change it to clearly say that in this run (for example) it happened in the 15th epoch and the Poutyne automatically saves the best weights (whether it is in 15 to 14, or...). I also used evaluate function to access predictions and expectations simultaneously.

Reviewer: I'm not sure it's the right thing to do because if someone run the notebook, it will not work and will have to manually change the filename in the code.

Author: I changed it to select checkpoint which contains the best weights without mentioning the epoch number directly. i think this file contains the best weights, doesn't it?

Reviewer: checkpoint.ckpt does not contain the best checkpoint but the last checkpoint. You should really use the load_checkpoint method in Experiment since it is there for that. You can pass it `'best'` to load the best checkpoint. Also, you can use the test method in Experiment instead of the evaluate_generator in Model to do the same thing you're currently doing. Or, you can get the Model inside the Experiment via `expt.model`.

The ``callbacks`` feature also records the training logs. we can use this information to monitor and analyze the training process.

.. code-block:: python

    logs = pd.read_csv('./callbacks/log.tsv', sep='\t')
    print(logs)

.. image:: /_static/img/classification_and_regression/logs.png

.. code-block:: python

    train_loss = logs.loss
    valid_loss = logs.val_loss
    plt.plot(train_loss)
    plt.plot(valid_loss)
    plt.legend(['train_loss','valid_loss'])
    plt.title('training and validation losses')
    plt.show()

.. image:: /_static/img/classification_and_regression/loss_diagram.png

We can also evaluate the performance of the trained network (a network with the best weights) on any dataset, as below:

.. code-block:: python

    exp.test(valid_dataloader)

Now let's evaluate the performance of the network visually.

.. code-block:: python

    sample_number = 35
    image = train_dataset[sample_number][0]
    image = image.permute(1,2,0).detach().numpy()
    image_rgb = cv2.cvtColor(np.float32(image), cv2.COLOR_BGR2RGB)
    image_rgb = image_rgb * imagenet_std + imagenet_mean

>     Gender = 'male' if np.argmax(predictions[0][sample_number])==0 else 'female'
>     print('Gender is: ', Gender)

Reviewer: ```suggestion
    gender = 'male' if np.argmax(predictions[0][sample_number])==0 else 'female'
    print('Gender is: ', gender)
```

.. code-block:: python

    data_root = 'datasets'
    base_url = "https://graal.ift.ulaval.ca/public/celeba/"

    file_list = [
        "img_align_celeba.zip",
        "list_attr_celeba.txt", 
        "identity_CelebA.txt",
        "list_bbox_celeba.txt",
        "list_landmarks_align_celeba.txt",
        "list_eval_partition.txt"
    ]

    # Path to folder with the dataset
    dataset_folder = f'{data_root}/celeba'
    os.makedirs(dataset_folder, exist_ok=True)

    for file in file_list:
        url = f"{base_url}/{file}"
        if not os.path.exists(f"{dataset_folder}/{file}"):
            wget.download(url, f"{dataset_folder}/{file}")

    with zipfile.ZipFile(f"{dataset_folder}/img_align_celeba.zip", 'r') as ziphandler:
        ziphandler.extractall(dataset_folder)
   
Now, as the dataset id downloaded, we can define our datasets and dataloaders in its original way.

.. code-block:: python

>     transforms = tfms.Compose ([
>         tfms.Resize((image_size, image_size)),
>         tfms.ToTensor(),
>         tfms.Normalize(imagenet_mean, imagenet_std)
>         ])

Reviewer: ```suggestion
    transforms = tfms.Compose([
        tfms.Resize((image_size, image_size)),
        tfms.ToTensor(),
        tfms.Normalize(imagenet_mean, imagenet_std)
    ])
```

>     train_dataset = datasets.CelebA(data_root,
>         split='train',
>         target_type=['attr', 'landmarks'],
>         transform=transforms)
>     valid_dataset = datasets.CelebA(data_root, 
>         split='valid', 
>         target_type=['attr', 'landmarks'], 
>         transform=transforms)
>     test_dataset = datasets.CelebA(data_root, 
>         split='test', 
>         target_type=['attr', 'landmarks'], 
>         transform=transforms)

Reviewer: ```suggestion
    train_dataset = datasets.CelebA(
        data_root,
        split='train',
        target_type=['attr', 'landmarks'],
        transform=transforms
    )
    valid_dataset = datasets.CelebA(
        data_root, 
        split='valid', 
        target_type=['attr', 'landmarks'], 
        transform=transforms)
    test_dataset = datasets.CelebA(
        data_root, 
        split='test', 
        target_type=['attr', 'landmarks'], 
        transform=transforms
    )
```

.. code-block:: python

    class ClassificationRegressionLoss(nn.Module):
        def __init__(self):
            super(ClassificationRegressionLoss, self).__init__()
            self.ce_loss = nn.CrossEntropyLoss() # size_average=False
            self.mse_loss = nn.MSELoss()
            
        def forward(self, y_pred, y_true):
            loss_cls = self.ce_loss(y_pred[0], Variable(y_true[0][:,20])) # Cross Entropy Error (for classification)
            loss_reg1 = self.mse_loss(y_pred[1][:,0], y_true[1][:,0]/h) # Mean Squared Error for X_L
            loss_reg2 = self.mse_loss(y_pred[1][:,1], y_true[1][:,1]/w) # Mean Squared Error for Y_L
            loss_reg3 = self.mse_loss(y_pred[1][:,2], y_true[1][:,2]/h) # Mean Squared Error for X_R
            loss_reg4 = self.mse_loss(y_pred[1][:,3], y_true[1][:,3]/w) # Mean Squared Error for Y_R

>             total_loss = loss_cls + W * (loss_reg1 + loss_reg2 + loss_reg3 + loss_reg4)

Reviewer: `W` should be parameter of the `__init__` method of the class, not a global as currently.

            wget.download(url, f"{dataset_folder}/{file}")

    with zipfile.ZipFile(f"{dataset_folder}/img_align_celeba.zip", 'r') as ziphandler:
        ziphandler.extractall(dataset_folder)
   
Now, as the dataset is downloaded, we can define our datasets and dataloaders in its original way.

.. code-block:: python

    transforms = tfms.Compose([
        tfms.Resize((image_size, image_size)),
        tfms.ToTensor(),
        tfms.Normalize(imagenet_mean, imagenet_std)
    ])
    train_dataset = datasets.CelebA(
        data_root,
        split='train',
        target_type=['attr', 'landmarks'],
        transform=transforms
    )
    valid_dataset = datasets.CelebA(
        data_root, 
        split='valid', 
        target_type=['attr', 'landmarks'], 
        transform=transforms
    )
    test_dataset = datasets.CelebA(
        data_root, 
        split='test', 
        target_type=['attr', 'landmarks'], 
        transform=transforms
    )
                                
    train_dataloader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    valid_dataloader = DataLoader(valid_dataset, batch_size=batch_size, shuffle=False)

>     test_dataloader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

Reviewer: You shouldn't create the dataloader here if you are creating them below. Move the test_dataloader creation with those below.

.. code-block:: python

    exp.load_checkpoint('best')
    model = exp.model
    loss, predictions, ground_truth = model.evaluate_generator(valid_dataloader, return_pred=True, return_ground_truth=True)

The ``callbacks`` feature of Poutyne, also used by the Experiment class, records the training logs. We can use this information to monitor and analyze the training process.

.. code-block:: python

    logs = pd.read_csv('./saves/two_task_example/log.tsv', sep='\t')
    print(logs)

.. image:: /_static/img/classification_and_regression/logs.png

.. code-block:: python

    train_loss = logs.loss
    valid_loss = logs.val_loss
    plt.plot(train_loss)
    plt.plot(valid_loss)
    plt.legend(['train_loss','valid_loss'])
    plt.title('training and validation losses')
    plt.show()

.. image:: /_static/img/classification_and_regression/loss_diagram.png

We can also evaluate the performance of the trained network (a network with the best weights) on any dataset, as below:

.. code-block:: python

    exp.test(test_dataloader)

Now let's evaluate the performance of the network visually.

.. code-block:: python

    sample_number = 35

>     image = valid_dataset[sample_number][0]

Reviewer: ```suggestion
    image = test_dataset[sample_number][0]
```
Use the test dataset here.

    optimizer = optim.Adam(network.parameters(), lr=learning_rate, weight_decay=0)
    loss_function = ClassificationRegressionLoss()
    #Step_Learning_Rate = StepLR(step_size=2 , gamma=0.1, last_epoch=-1, verbose=False)
    exp = Experiment('./two_task_example', network, optimizer=optimizer, loss_function=loss_function, device="all")
    exp.train(train_dataloader, valid_dataloader, callbacks=callbacks, epochs=num_epochs)

Evaluation
==========

As you have also noticed from the training logs, in this try we achieved the best performance (considering the validation loss) at the 15th epoch. The weights of the network for the corresponding epoch have been automatically saved by the `Experiment` function and we use these parameters to evaluate our algorithm visually. For this purpose, we utulize the load_checkpoint method and set its argument to `best` to load the best weights of the model automatically. Finally,  we take advantage of the `evaluate` function of Poutyne, and apply it to the validation dataset. It provides us the predictions as well as the ground-truth for comparison, in case of need.

.. code-block:: python

    exp.load_checkpoint('best')
    model = exp.model

>     loss, predictions, Ground_Truth = model.evaluate_generator(valid_dataloader, return_pred=True, return_ground_truth=True)

Reviewer: ```suggestion
    loss, predictions, ground_truth = model.evaluate_generator(valid_dataloader, return_pred=True, return_ground_truth=True)
```

.. code-block:: python

    data_root = 'datasets'
    base_url = "https://graal.ift.ulaval.ca/public/celeba/"

    file_list = [
        "img_align_celeba.zip",
        "list_attr_celeba.txt", 
        "identity_CelebA.txt",
        "list_bbox_celeba.txt",
        "list_landmarks_align_celeba.txt",
        "list_eval_partition.txt"
    ]

    # Path to folder with the dataset
    dataset_folder = f'{data_root}/celeba'
    os.makedirs(dataset_folder, exist_ok=True)

    for file in file_list:
        url = f"{base_url}/{file}"
        if not os.path.exists(f"{dataset_folder}/{file}"):
            wget.download(url, f"{dataset_folder}/{file}")

    with zipfile.ZipFile(f"{dataset_folder}/img_align_celeba.zip", 'r') as ziphandler:
        ziphandler.extractall(dataset_folder)
   

> Now, as the dataset id downloaded, we can define our datasets and dataloaders in its original way.

Reviewer: ```suggestion
Now, as the dataset is downloaded, we can define our datasets and dataloaders in its original way.
```

.. code-block:: python

    transforms = tfms.Compose([
        tfms.Resize((image_size, image_size)),
        tfms.ToTensor(),
        tfms.Normalize(imagenet_mean, imagenet_std)
    ])
    train_dataset = datasets.CelebA(
        data_root,
        split='train',
        target_type=['attr', 'landmarks'],
        transform=transforms
    )
    valid_dataset = datasets.CelebA(
        data_root, 
        split='valid', 
        target_type=['attr', 'landmarks'], 
        transform=transforms
    )
    test_dataset = datasets.CelebA(
        data_root, 
        split='test', 
        target_type=['attr', 'landmarks'], 
        transform=transforms
    )
                                
    train_dataloader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    valid_dataloader = DataLoader(valid_dataset, batch_size=batch_size, shuffle=False)
    test_dataloader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

Here we can see how each dataset sample looks like:

.. code-block:: python

>     print (train_dataset[0])

Reviewer: An image of the output could be added here.

Author: <img width="809" alt="output" src="https://user-images.githubusercontent.com/50740990/128584636-a0572603-589e-443f-bd6c-b00323e75778.png">
 like this?

Reviewer: Yeah, looks good.

.. code-block:: python

    class ClassificationRegressionLoss(nn.Module):
        def __init__(self, W):
            super(ClassificationRegressionLoss, self).__init__()
            self.ce_loss = nn.CrossEntropyLoss() # size_average=False
            self.mse_loss = nn.MSELoss()
            self.W = W
            
        def forward(self, y_pred, y_true):
            loss_cls = self.ce_loss(y_pred[0], Variable(y_true[0][:,gender_index])) # Cross Entropy Error (for classification)
            loss_reg1 = self.mse_loss(y_pred[1][:,0], y_true[1][:,0]/h) # Mean Squared Error for X_L
            loss_reg2 = self.mse_loss(y_pred[1][:,1], y_true[1][:,1]/w) # Mean Squared Error for Y_L
            loss_reg3 = self.mse_loss(y_pred[1][:,2], y_true[1][:,2]/h) # Mean Squared Error for X_R
            loss_reg4 = self.mse_loss(y_pred[1][:,3], y_true[1][:,3]/w) # Mean Squared Error for Y_R
            total_loss = loss_cls + self.W * (loss_reg1 + loss_reg2 + loss_reg3 + loss_reg4)
            return total_loss

Training
========

.. code-block:: python

    optimizer = optim.Adam(network.parameters(), lr=learning_rate, weight_decay=0)

>     loss_function = ClassificationRegressionLoss()

Reviewer: ```suggestion
    loss_function = ClassificationRegressionLoss(W)
```

    #Step_Learning_Rate = StepLR(step_size=2 , gamma=0.1, last_epoch=-1, verbose=False)

>     exp = Experiment('./two_task_example', network, optimizer=optimizer, loss_function=loss_function, device="all")

Reviewer: ```suggestion
    exp = Experiment('./saves/two_task_example', network, optimizer=optimizer, loss_function=loss_function, device="all")
```

    exp.train(train_dataloader, valid_dataloader, callbacks=callbacks, epochs=num_epochs)

Evaluation
==========

> As you have also noticed from the training logs, in this try we achieved the best performance (considering the validation loss) at the 15th epoch. The weights of the network for the corresponding epoch have been automatically saved by the `Experiment` function and we use these parameters to evaluate our algorithm visually. For this purpose, we utulize the load_checkpoint method and set its argument to `best` to load the best weights of the model automatically. Finally,  we take advantage of the `evaluate` function of Poutyne, and apply it to the validation dataset. It provides us the predictions as well as the ground-truth for comparison, in case of need.

Reviewer: ```suggestion
As you have also noticed from the training logs, in this try we achieved the best performance (considering the validation loss) at the 15th epoch. The weights of the network for the corresponding epoch have been automatically saved by the `Experiment` function and we use these parameters to evaluate our algorithm visually. For this purpose, we utilize the `load_checkpoint` method and set its argument to `best` to load the best weights of the model automatically. Finally,  we take advantage of the `evaluate` function of Poutyne, and apply it to the validation dataset. It provides us the predictions as well as the ground-truth for comparison, in case of need.
```

.. code-block:: python

    exp.load_checkpoint('best')
    model = exp.model
    loss, predictions, ground_Truth = model.evaluate_generator(valid_dataloader, return_pred=True, return_ground_truth=True)

> The ``callbacks`` feature also records the training logs. we can use this information to monitor and analyze the training process.

Reviewer: ```suggestion
The ``callbacks`` feature of Poutyne, also used by the Experiment class, records the training logs. We can use this information to monitor and analyze the training process.
```

.. code-block:: python

>     logs = pd.read_csv('./two_task_example/log.tsv', sep='\t')

Reviewer: ```suggestion
    logs = pd.read_csv('./saves/two_task_example/log.tsv', sep='\t')
```

    loss_function = ClassificationRegressionLoss(W)
    #Step_Learning_Rate = StepLR(step_size=2 , gamma=0.1, last_epoch=-1, verbose=False)
    exp = Experiment('./saves/two_task_example', network, optimizer=optimizer, loss_function=loss_function, device="all")
    exp.train(train_dataloader, valid_dataloader, callbacks=callbacks, epochs=num_epochs)

Evaluation
==========

As you have also noticed from the training logs, in this try we achieved the best performance (considering the validation loss) at the 15th epoch. The weights of the network for the corresponding epoch have been automatically saved by the `Experiment` function and we use these parameters to evaluate our algorithm visually. For this purpose, we utilize the `load_checkpoint` method and set its argument to `best` to load the best weights of the model automatically. Finally,  we take advantage of the `evaluate` function of Poutyne, and apply it to the validation dataset. It provides us the predictions as well as the ground-truth for comparison, in case of need.

.. code-block:: python

    exp.load_checkpoint('best')
    model = exp.model

>     loss, predictions, ground_Truth = model.evaluate_generator(valid_dataloader, return_pred=True, return_ground_truth=True)

Reviewer: ```suggestion
    loss, predictions, ground_truth = model.evaluate_generator(valid_dataloader, return_pred=True, return_ground_truth=True)
```

.. role:: hidden
    :class: hidden-section

Gender Classification and Eyes Location Detection: A Two Task Problem
*********************************************************************

.. note::

    - See the notebook `here <https://github.com/GRAAL-Research/poutyne/blob/master/examples/classification_and_regression.ipynb>`_
    - Run in `Google Colab <https://colab.research.google.com/github/GRAAL-Research/poutyne/blob/master/examples/classification_and_regression.ipynb>`_

In this example, we are going to implement a multi-task problem. We try to identify the gender of the people, as well as locating their eyes in the image. Hence, we have two different tasks: classification (to identify the gender) and regression (to find the location of the eyes). We are going to use a single network (A CNN) to perform both tasks, however, we will need to apply different loss functions, each proper to a specific task.

Let's import all the needed packages.

.. code-block:: python

    import math
    import os
    import matplotlib.pyplot as plt
    import numpy as np
    import pandas as pd
    import gdown
    import zipfile
    import cv2

>     from natsort import natsorted

Reviewer: What packages do we need to install for these two?

Author: pip install natsort

Reviewer: The text should specify that they are needed since they are not usually installed.

Reviewer: This was not done in the last commit.

Author: I added a text in the IPYNP as I included the codes to install them. in previous examples, you mentioned removing installation codes ( pip install poutyne for example) from the rest file. 
So in the rest file do you want me to just mention the need for installation of also to include the installation codes?

Reviewer: Just for these since they are unusual, add a line in the RST saying you use these libraries and add the lines to install them.

Fetching data 
=============

The section below consists of a few lines of codes that help us download celebA dataset from a public web source and unzip them.

.. code-block:: python

    data_root = 'data/celeba'
    # Path to folder with the dataset
    dataset_folder = f'{data_root}/img_align_celeba'

    # URL for the CelebA dataset (aligned images, attributes, landmasrks)
    url = 'https://graal.ift.ulaval.ca/public/celeba/img_align_celeba.zip'
    attr_url = 'https://graal.ift.ulaval.ca/public/celeba/list_attr_celeba.txt'
    land_mark_url = 'https://graal.ift.ulaval.ca/public/celeba/list_landmarks_align_celeba.txt'

    # Path to download the dataset to
    download_path = f'{data_root}/img_align_celeba.zip'
    land_mark_path = f'{data_root}/list_landmarks_align_celeba.txt'
    attr_path = f'{data_root}/list_attr_celeba.txt'

    # Create required directories 

>     if not os.path.exists(data_root):

Reviewer: Use `exists_ok=True` with `os.makedirs` instead.

    os.makedirs(data_root)
    os.makedirs(dataset_folder)

    # Download the dataset from the source
    gdown.download(url, download_path, quiet=False)
    gdown.download(land_mark_url, land_mark_path, quiet=False)

>     gdown.download(attr_url, attr_path, quiet=False)

Reviewer: You don't need to use gdown since the links are not on Google Drive.

Author: i just found it easy for downloading. what do you recommend? 
how about "request.get" ?

Reviewer: Yes, requests is good.

    # Unzip the downloaded file 
    with zipfile.ZipFile(download_path, 'r') as ziphandler:

>     ziphandler.extractall(dataset_folder)

Reviewer: Missing indentation spaces.

   
Create a custom dataset class
=============================

As we are going to implement a multi-task problem by a single CNN, we should provide the CNN with the ground truth in a proper way. Here, we have two different tasks: classification and regression. In the classification task, the goal is to identify the gender. The labels of the gender for each image are saved in the `list_attr_celeba.txt` file, in which 1 stands for male and -1 for female. Since we consider the loss of both tasks simultaneously, we scale all target values to the range of [0,1]. Hence, the gender labels will be changed as well, 1 for male and 0 for female. For the localization part, the coordinates of the eyes (Left and Right) are provided in the `list_landmarks_align_celeba.txt` file. In addition to scaling the number to the range of [0,1], we also need to rescale the coordinates to the image's new size (224,224).

.. code-block:: python

>     class CelebADataset(Dataset):

Reviewer: Why don't you use [torchvision's CelabA dataset](https://pytorch.org/vision/stable/datasets.html#celeba) instead?

Reviewer: @mohammad-brdrn You can download the dataset like you do before instantiating the torchvision dataset. In that way, it won't try to download it from Internet.

Author: you mean using torchvision.datasets.celebA(root=root, download=True)?
If so, that's the problem I mentioned. it always fails to download. also, torch communities also admit that and suggest downloading from the internet. for example link below:

https://github.com/pytorch/vision/issues/2262

Reviewer: Documentation says:

> download (bool, optional) – If true, downloads the dataset from the internet and puts it in root directory. If dataset is already downloaded, it is not downloaded again.

So if you download the dataset via the current links and put the files under the file hierarchy expected by torchvision.datasets.CelebA, it will work.

Author: I put several days on this trying different strategies but still, we cannot use torch.datasets.CelbA as the standard dataset. because:

1)we publish this example mainly on google collab for the readers. Googlecolab does not keep downoladed files after runtime. so the problem still exists.
2) I tried downloading the file (first by setting download=true) and using it with (download=False). I tried it more than 9 days at almost any time (even midnight) to gives the same error.
3) I read almost all relevant posts on this issue and also inspected the root after the error. in fact, after the error (BadZipFile: File is not a zip file) i saw that it has downloaded all the files!! but it is not able to unzip the file! i tried downloading and unzipping the zipfile with other ways, but it doesnt work again.
4) can we proceed with the proposed dataset class? in meanwhile i will try other ways as well. i don't want to stop other parts for this long time, focusing on this part. maybe i can resume it along with other parts.

<img width="444" alt="3afterzipls" src="https://user-images.githubusercontent.com/50740990/119734646-48e83680-be49-11eb-8734-fd9b3e525c3b.png">
<img width="491" alt="1datadown" src="https://user-images.githubusercontent.com/50740990/119734731-60bfba80-be49-11eb-9ddc-c5635f96e5b4.png">
<img width="391" alt="2zip" src="https://user-images.githubusercontent.com/50740990/119734749-6a492280-be49-11eb-9a73-bb75a4f447ee.png">

Reviewer: Hi @mohammad-brdrn , sorry for the late response. Here is some code which works for using the CelebA dataset:

For loading the files:
```python
data_root = 'datasets'

base_url = "https://graal.ift.ulaval.ca/public/celeba/"

file_list = [
    "img_align_celeba.zip",
    "list_attr_celeba.txt", 
    "identity_CelebA.txt",
    "list_bbox_celeba.txt",
    "list_landmarks_align_celeba.txt",
    "list_eval_partition.txt"
]

# Path to folder with the dataset
dataset_folder = f'{data_root}/celeba'
os.makedirs(dataset_folder, exist_ok=True)

for file in file_list:
    url = f"{base_url}/{file}"
    wget.download(url, f"{dataset_folder}/{file}")

with zipfile.ZipFile(f"{dataset_folder}/img_align_celeba.zip", 'r') as ziphandler:
    ziphandler.extractall(dataset_folder)
```
Then, instantiating the datasets:
```python
train_dataset = datasets.CelebA(data_root, 
                                split='train', 
                                target_type=['attr', 'landmarks'], 
                                transform=tfms.ToTensor())
valid_dataset = datasets.CelebA(data_root, 
                                split='valid', 
                                target_type=['attr', 'landmarks'], 
                                transform=tfms.ToTensor())
test_dataset = datasets.CelebA(data_root, 
                               split='test', 
                               target_type=['attr', 'landmarks'], 
                               transform=tfms.ToTensor())
```

You had to look in the [code of CelebA](https://pytorch.org/vision/stable/_modules/torchvision/datasets/celeba.html#CelebA) to know how to do it.

Reviewer: Btw, when using wget, you should verify that the files do not exist before downloading. Otherwise, it creates files with a suffix "(1)".

Author: I applied the changes ( using torch.datasets.CelebA), While you check the overall format, I may work more on hyperparameters.

    valid_indices = indices[:math.floor(full_dataset_length * valid_split_percent)]
    train_dataset = Subset(celeba_dataset, train_indices)
    valid_dataset = Subset(celeba_dataset, valid_indices)
    train_dataloader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    valid_dataloader = DataLoader(valid_dataset, batch_size=batch_size, shuffle=False)

Here we can see how each dataset sample looks like:

.. code-block:: python

    print (train_dataset[0])

Here, we can see an example from the training dataset. It shows an image of a person, printing the gender and also showing the location of the eyes.

.. code-block:: python

    sample_number = 16
    image = train_dataset[sample_number][0]
    image = image.permute(1,2,0).detach().numpy()
    image_rgb = cv2.cvtColor(np.float32(image), cv2.COLOR_BGR2RGB)
    Gender = 'male' if int(train_dataset[sample_number][1][0])==1 else 'female'
    print('Gender is: ', Gender)
    w, h = train_dataset[sample_number][1][2]
    (x1, y1) = train_dataset[sample_number][1][1][0:2]
    (x2, y2) = train_dataset[sample_number][1][1][2:4]
    x1, x2 = int(x1*h), int(x2*h)
    y1, y2 = int(y1*w), int(y2*w)

>     image_rgb	=	cv2.drawMarker(image_rgb, (x1,y1), (0,255,0))
>     image_rgb	=	cv2.drawMarker(image_rgb, (x2,y2), (0,255,0))

Reviewer: ```suggestion
    image_rgb = cv2.drawMarker(image_rgb, (x1,y1), (0,255,0))
    image_rgb = cv2.drawMarker(image_rgb, (x2,y2), (0,255,0))
```

    image_rgb = np.clip(image_rgb , 0, 1)
    plt.imshow(image_rgb)
    plt.axis('off')

>     plt.show()

Reviewer: The colors of the output are a little off.

Author: done

.. image:: /_static/img/classification_and_regression/dataset_sample.png

Network
=======

Below, we define a new class, named 'ClassifierLocalizer, which accepts a pre-trained CNN and changes its last fully connected layer to be proper for the two task problem. The new fully connected layer contains 6 neurons, 2 for the classification task (male or female) and 4 for the localization task (x and y for the left and right eyes). Moreover, to put the location results on the same scale as the class scores, we apply the sigmoid function to the neurons assigned for the localization task.

.. code-block:: python

    class ClassifierLocalizer(nn.Module):
        def __init__(self, model_name, num_classes=2):
            super(ClassifierLocalizer, self).__init__()
            self.num_classes = num_classes
            
            # create cnn model

>             model = models.__dict__[model_name](True)

Reviewer: Prefer using `getattr` instead of `__dict__`.

Author: i'll change it

Author: done

Regarding the complexity of the problem, the number of the samples in the training dataset, and the similarity of the training dataset to the ImageNet dataset, we may decide to freeze some of the layers. In our current example, based on the mentioned factors, we freeze just the last fully connected layer.

.. code-block:: python

    network = ClassifierLocalizer(model_name='resnet18')

    def freeze_weights(network):
        for name, param in network.named_parameters():
            if not name.startswith('fc.'):
                param.requires_grad = False

    freeze_weights(network)
    print(network)

.. code-block:: python

    network = ClassifierLocalizer(model_name='resnet18')  # network without freezing any layer.

Loss function
=============

As we discussed before, we have two different tasks in this example. These tasks need different loss functions; Cross-Entropy loss for the classification and Mean Square Error loss for the regression. Below, we define a new loss function class that sums both losses to considers them simultaneously. However, as the regression is relatively a harder task, we apply a higher weight to MSEloss.

.. code-block:: python

>     class classification_regression_Loss(nn.Module):

Reviewer: Should be in Pascal case, that is `ClassificationRegressionLoss`

        def __init__(self):
            super(classification_regression_Loss, self).__init__()
            self.ce_loss = nn.CrossEntropyLoss() # size_average=False
            self.mse_loss = nn.MSELoss()
            
        def forward(self, y_pred, y_true):
            loss_cls = self.ce_loss(y_pred[0], y_true[0]) # Cross Entropy Error (for classification)
            loss_reg = self.mse_loss(y_pred[1], y_true[ 1]) # Mean Squared Error (for landmarks)
            total_loss = loss_reg + loss_cls
            return total_loss

Training
========

.. code-block:: python

    optimizer = optim.Adam(network.parameters(), lr=0.0001, weight_decay=0)
    loss_function = classification_regression_Loss()
    #Step_Learning_Rate = StepLR(step_size=2 , gamma=0.1, last_epoch=-1, verbose=False)
    exp = Experiment('./two_task_example', network, optimizer=optimizer, loss_function=loss_function, device="all")

>     exp.train(train_dataloader, valid_dataloader, callbacks=callbacks, epochs=num_epochs)

Reviewer: There is no `callbacks` defined above. Also, the ones you defined in the notebook are not useful since you are using Experiment and Experiment already has all of them.

Author: I agree with you. however in the code:
**exp.train(train_dataloader, valid_dataloader, callbacks=callbacks, epochs=num_epochs)**
if i want to use the callbacks argument, i should define it first. isn't it?

Reviewer: I don't understand, what callbacks do you want to use?

Author: Solved

    plt.show()

.. image:: /_static/img/classification_and_regression/loss_diagram.png

We can also evaluate the performance of the trained network (a network with the best weights) on any dataset, as below:

.. code-block:: python

    exp.test(valid_dataloader)

Now let's evaluate the performance of the network visually.

.. code-block:: python

    sample_number = 10
    image = valid_dataset[sample_number][0]
    image = image.permute(1,2,0).detach().numpy()
    image_rgb = cv2.cvtColor(np.float32(image), cv2.COLOR_BGR2RGB)
    #Gender = 'male' if np.where(predictions[0][sample_number]==max(predictions[0][sample_number]))[0]==0 else 'female'
    Gender = 'male' if np.argmax(predictions[0][sample_number])==0 else 'female'
    print('Gender is: ', Gender)
    w, h = valid_dataset[sample_number][1][2]
    (x1, y1) = predictions[1][sample_number][0:2]
    (x2, y2) = predictions[1][sample_number][2:4]
    x1, x2 = int(x1*h), int(x2*h)
    y1, y2 = int(y1*w), int(y2*w)
    image_rgb	=	cv2.drawMarker(image_rgb, (x1,y1), (0,255,0))
    image_rgb	=	cv2.drawMarker(image_rgb, (x2,y2), (0,255,0))
    image_rgb = np.clip(image_rgb , 0, 1)
    plt.imshow(image_rgb)
    plt.axis('off')

>     plt.show()

Reviewer: Same comments as above.
